app/src/androidTest/java/ch/epfl/sweng/SDP/auth/LoginActivityTest.java

```diff
@@ -37,51 +47,99 @@
-                .thenReturn(mockIdpResponse);
         Mockito.when(mockIdpResponse.isNewUser()).thenReturn(true);
-        loginActivity.onActivityResult(42, -1, mockIntent);
+        activityRule.getActivity().onActivityResult(42, -1, mockIntent);
         assertTrue(loginActivity.isFinishing());
         Activity accountCreationActivity = getInstrumentation()
                 .waitForMonitorWithTimeout(monitor, 5000);
         assertThat(accountCreationActivity, is(not(nullValue())));
     }
 
     @Test
-    public void testCloneAccountFromFirebase() {
-        DataSnapshot mockSnapshot = Mockito.mock(DataSnapshot.class);
-        HashMap<String, HashMap<String, Object>> userEntry = new HashMap<>();
-        Mockito.when(mockSnapshot.getValue())
-                .thenReturn(userEntry);
-        assertThat(mockSnapshot.getValue(), CoreMatchers.<Object>is(userEntry));
+    public void testSuccessfulLoginExistingUser() {
+        Mockito.when(mockIdpResponse.isNewUser()).thenReturn(false);
+        activityRule.getActivity().onActivityResult(42, -1, mockIntent);
+        SystemClock.sleep(3000);
+        assertTrue(loginActivity.isFinishing());
+        Activity homeActivity = getInstrumentation()
+                .waitForMonitorWithTimeout(homeMonitor, 5000);
+        Assert.assertNotNull(homeActivity);
+    }
+
+    @Test
+    public void testSuccessfulLoginWithoutAccount() {
+        Mockito.when(mockIdpResponse.getEmail()).thenReturn("weirdEmail");
+        Mockito.when(mockIdpResponse.isNewUser()).thenReturn(false);
+        activityRule.getActivity().onActivityResult(42, -1, mockIntent);
+        Activity accountCreationActivity = getInstrumentation()
+                .waitForMonitorWithTimeout(monitor, 5000);
+        Assert.assertNotNull(accountCreationActivity);
+    }
+
+    @Test
+    public void testFailedLoginNoNetwork() {
+        failedSignInHelper(R.string.no_internet, ErrorCodes.NO_NETWORK);
+    }
+
+    @Test
+    public void testFailedLoginUnknownError() {
+        failedSignInHelper(R.string.unknown_error, 1234567891);
+    }
+
+    private void failedSignInHelper(int expectedErrorMessageId, final int errorCode) {
+        executeOnUiThread(new Runnable() {
+            @Override
+            public void run() {
+                activityRule.getActivity().handleFailedSignIn(errorCode);
+            }
+        });
+
+        TextView feedbackView = loginActivity.findViewById(R.id.error_message);
+        ViewMatchers.assertThat(feedbackView.getText().toString(), is(equalTo(
+               loginActivity.getResources().getString(expectedErrorMessageId))));
+    }
+
+    private void executeOnUiThread(Runnable runnable) {
+        try {
+            runOnUiThread(runnable);
+        } catch (Throwable throwable) {
+            throwable.printStackTrace();
+        }
     }
 
     @Test
     public void testPressingBackButtonDoesNothing() {
         activityRule.getActivity().onBackPressed();
```

Comment:
Why no assertion here?

Reply:
Not your code but still

Reply:
You are right. However, as you can see, I didn't modify this part of the code. But I will change that.

Reply:
👍